### PR TITLE
Support bundle

### PR DESCRIPTION
**Pull request by mpfz0r**
## Support Bundle

Add the possibility to export a zip file which contains often needed debugging information from a Graylog cluster.

### Bundle Content

 - Logs from all cluster nodes
    - Current log from RollingFileAppender and older, rotated logs
    - In-Memory log. Useful for docker environments, where there are no log files in the filesystem.

 - Version Information from all nodes
   - [x] Graylog version
   - [x] is_leader status
   - [x] Installed Plugins
   - [x] Elastic Version
   - [x] Mongo Version
   - [x] OS version
   - [x] OSHI stats
   - [x] System metrics

### Implementation notes

- Build a node endpoint to fetch a `SupportBundleNodeManifest` containing
    - All available log files
  
The `SupportBundleNodeManifest` structure:

```json
{
  "entries": {
    "logfiles": [
      { "id":  "0",      "filename":  "/tmp/server.log", "size":  7584  },
      { "id":  "1",      "filename":  "/tmp/server.log.1.gz", "size":  17589 },
      { "id":  "memory", "filename":  "server.mem.log", "size":  -1 }
    ]
  }
}

```

- Build a node endpoint to fetch individual log files from each cluster node.
      This should be a `StreamingOutput`

- Build a cluster endpoint to create a new support bundle.
  ~~This will create a system job that should be restricted to the current leader.~~ (postponed. likely fine without)
- Implement a ~~system job~~ service that creates the support bundle.
  - It needs to fetch all individual logs, stats etc. from all nodes into a tmp directory. (`${data_dir}/support-bundle`)
  File structure:
```text
data_dir/
  support-bundle/
    support-bundle-2023-03-14T10:12:23.zip
    .tmp.bundle-build-dir-34824/
        cluster.json
        a107430c/
          thread-dump.txt
          logs/
             server.0.log
             server.1.log.gz
             server.mem.log
        b42c94d0/
          thread-dump.txt
          logs/
             server.0.log
             server.1.log.gz
             server.2.log.gz
             server.mem.log
       

```
  - build a zip file (timestamp in filename) with all that content. and remove tmp dir.

- Build a leader-only (proxying) resource that serves the zip bundle files.
- Build a leader-only (proxying) resource that allows deletion of the zip bundle files.

**UI:**
Screenshot 2023-03-23 at 16 32 27

/jenkins-pr-deps https://github.com/Graylog2/graylog-plugin-enterprise/pull/4884

### CODE REVIEW

**Comment by patrickmann** [comment]
Building the bundle seems complex enough to be worth some unit tests.

**Review by patrickmann — APPROVED** [review]
Performs as expected

File: graylog2-web-interface/src/components/loggers/ClusterSupportBundleOverview.tsx
Comment (ousmaneo) [inline graylog2-web-interface/src/components/loggers/ClusterSupportBundleOverview.tsx:1]: Can we add a test for this component?

File: graylog2-web-interface/src/components/loggers/ClusterSupportBundleOverview.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+              </colgroup>
+              <thead>
+                <tr>
+                  <th>Filename</th>
+                  <AlignRightHeadCell>Size</AlignRightHeadCell>
+                </tr>
+              </thead>
+              <tbody>
+                {list.map((bundle) => (
+                  <tr key={bundle.file_name}>
```
Comment (ousmaneo) [inline graylog2-web-interface/src/components/loggers/ClusterSupportBundleOverview.tsx:82]: I think we should extract the row rendering to a component for better visibility.

File: graylog2-web-interface/src/hooks/useClusterSupportBundle.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  } catch (errorThrown) {
+    UserNotification.error(`Downloading the Support Bundle failed with status: ${errorThrown}`, 'Could not download the Support Bundle.');
+  }
+};
+
+const useClusterSupportBundle = () => {
+  const [isCreating, setIsCreating] = useState<boolean>(false);
+  const { data, refetch } = useQuery(
+    ['supportBundleList', 'overview'],
+    fetchSupportBundleList,
```
Comment (ousmaneo) [inline graylog2-web-interface/src/hooks/useClusterSupportBundle.ts:63]: Can we add types for the `data` or use 'graylog-api' to get types from the API?

**Review by ousmaneo — CHANGES_REQUESTED** [review]
I started reviewing this PR and I added some comments inline. I will test now.

**Review by ousmaneo — APPROVED** [review]
LGTM!

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleClusterResource.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                return response.build();
+            } catch (Exception e) {
+                responseBody.close();
+            }
+        }
+        return Response.status(nodeResponse.code()).entity(nodeResponse.body()).build();
+    }
+
+    @DELETE
+    @Path("/bundle/delete/{filename}")
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleClusterResource.java:144]: yeah, I agree :+1:

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleService.java [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+        try (var files = Files.walk(bundleDir)) {
+            return files
+                    .filter(p -> !Files.isDirectory(p))
+                    .filter(p -> p.getFileName().toString().startsWith(BUNDLE_NAME_PREFIX))
+                    .map(f -> {
+                        try {
+                            return new BundleFile(f.getFileName().toString(), Files.size(f));
+                        } catch (IOException e) {
+                            LOG.warn("Exception while trying to list support bundles", e);
+                            throw new BadRequestException(e);
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleService.java:497]: Not really due to a "bad request". See the comment above.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleService.java [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+                        }
+                    })
+                    .sorted(Comparator.comparing(BundleFile::fileName).reversed())
+                    .collect(Collectors.toList());
+        } catch (NoSuchFileException ignored) {
+            return List.of();
+        } catch (IOException e) {
+            LOG.warn("Exception while trying to list support bundles", e);
+            throw new BadRequestException(e);
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/debug/bundle/SupportBundleService.java:506]: Not really due to a "bad request". See the comment above.

File: graylog2-server/src/main/resources/log4j2.xml [hunk 6]
```diff
@@ -32,6 +35,7 @@
         <Logger name="com.azure.core.amqp.implementation.ReactorReceiver" level="off"/>
         <Logger name="com.azure.core.amqp.implementation.ReactorDispatcher" level="off"/>
         <Root level="warn">
+            <AppenderRef ref="rolling-file"/>
```
Comment (thll) [inline graylog2-server/src/main/resources/log4j2.xml:38]: Does this belong here?

Comment (mpfz0r) [reply graylog2-server/src/main/resources/log4j2.xml:38]: oops, missed to delete that

File: graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * Write the buffer content into the provided OutputStream.
+     * @param outputStream The OutputStream to write the entries into.
+     * @param limit    limit the returned entries. This is only a rough estimate and will be rounded to the nearest batch size.
+     */
+    public synchronized void streamContent(OutputStream outputStream, int limit) {
+        // Round limit to the nearest batch size
+        final AtomicInteger count = new AtomicInteger(limit == 0 ? Integer.MAX_VALUE : limit / BATCHSIZE + 1);
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java:108]: yeah, that can be improved. actually I found a bug here.. we were not returning the newest entries from the compressed buffer, but the oldest ones :facepalm:

File: graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java [hunk 8]
```diff
@@ -104,24 +103,36 @@ private void flush() throws IOException {
...
 
         compressedRingBuffer.stream().map(input -> {
                     try {
                         return decompress(input);
                     } catch (IOException e) {
                         throw new RuntimeException(e);
                     }
-                }).takeWhile(b -> count.getAndDecrement() > 1)
+                })
+                .skip(skipFromCompressed)
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java]: I think we can move the skip ahead of the decompression.

**Review by thll — APPROVED** [review]
One comment, but approving anyways.

File: graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java (hunk 8 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java]: absolutely :+1:

**Pull request merged**